aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -571,95 +571,106 @@ Tensor sparse_compressed_to_dense(
+      nrows = self.size(batch_ndim) / blocksize[0];
+      ncols = self.size(batch_ndim + 1) / blocksize[1];
+      dense_reshaped_sizes[1] = blocksize[0];
+      dense_reshaped_sizes[2] = blocksize[1];
     }
-    // Due to early exit above this reshape should always be valid
-    dense = dense.reshape({n_batch, -1, values.size(-2), values.size(-1)});
-    for (auto batch : c10::irange(n_batch)) {
-      Tensor batch_indices = at::_convert_indices_from_csr_to_coo(
-          compressed_indices[batch],
-          plain_indices[batch],
-          false,
-          self.layout() == kSparseBsc);
-      auto batch_row_indices = batch_indices.select(0, 0);
-      auto batch_col_indices = batch_indices.select(0, 1);
-      auto offsets = batch_col_indices +
-          batch_row_indices * (self.size(-1) / blocksize[1]);
-      dense[batch].index_add_(0, offsets, values[batch]);
+    dense_reshaped_sizes[0] = n_batch * nrows * ncols;
+    dense = dense.reshape(dense_reshaped_sizes);
+
+    // Calculate batch, row and column indices for non-zeros in the
+    // sparse matrix, and use these to calculate correspoding indices
+    // into the dense matrix reshaped as above.  Then, update dense
+    // matrix by adding sparse matrix values into elements with
+    // indices calculated this way.
+    auto options = compressed_indices.options();
+    auto nnz_per_batch = values.size(1);
+    auto batch_indices = at::arange(0, n_batch, options).repeat_interleave(nnz_per_batch);
+    auto ncompressed = compressed_rows ? nrows : ncols;
+    auto compressed_indices_over_all_batches =
+      at::cat({compressed_indices.slice(1, 0, ncompressed).flatten()
+              + nnz_per_batch * at::arange(0, n_batch, options).repeat_interleave(ncompressed),
+              n_batch * nnz_per_batch * at::ones({1}, options)});
+    Tensor indices = at::_convert_indices_from_csr_to_coo(
+        compressed_indices_over_all_batches,
+        plain_indices.flatten(),
+        false,
+        !compressed_rows);
+    auto row_indices = indices.select(0, 0);
+    auto col_indices = indices.select(0, 1);
+    if (compressed_rows) {
+      row_indices -= batch_indices * nrows;
+    } else {
+      col_indices -= batch_indices * ncols;
     }
+    auto offsets = col_indices + row_indices * ncols + batch_indices * nrows * ncols;
+    dense.index_add_(0, offsets, values.flatten(0, 1));
 
-    // un-tile the result, NOTE: The final reshape uses the original
-    // self.sizes() which will squeeze out the extra batch dim if we put one in
-    return dense
-        .unflatten(
-            1, {self.size(-2) / blocksize[0], self.size(-1) / blocksize[1]})
+    // Un-tile the result.  The final reshape uses the original
+    // self.sizes() which will squeeze out the extra batch dim if we
+    // put one in.
+    if (!block_sparse) {
+      return dense.reshape(self.sizes());
+    } else {
+      return dense
+        .unflatten(0, {-1, nrows, ncols})
         .transpose(2, 3)
```

Comment:
Yes, it is correct.  

For example, for:
```python
crow_indices = [0, 1, 2]
col_indices = [0, 1]
values = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
x = torch.sparse_bsr_tensor(torch.tensor(crow_indices, dtype=torch.int64), torch.tensor(col_indices, dtype=torch.int64), torch.tensor(values), dtype=torch.double)
x.to_dense()
```
it produces:
```python
tensor([[1., 2., 0., 0.],
        [3., 4., 0., 0.],
        [0., 0., 5., 6.],
        [0., 0., 7., 8.]], dtype=torch.float64)
```
as expected, while without the transpose it would give:
```python
tensor([[1., 2., 3., 4.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [5., 6., 7., 8.]], dtype=torch.float64)
```
that is wrong.